tiledb/sm/c_api/tiledb.h

```diff
@@ -3201,6 +3234,468 @@ TILEDB_EXPORT int32_t tiledb_array_encryption_type(
+ * tiledb_array_open(ctx, array, TILEDB_READ);
+ * tiledb_subarray_t* subarray;
+ * tiledb_subarray_alloc(ctx, array, TILEDB_UNORDERED, &subarray);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param array The array that the subarray corresponds to.
+ * @param layout The layout of either the results if the subarray is
+ *     used for reading, or the layout of the values that the user will provide
+ *     for writing.
+ * @param subarray The subarray object to be created.
+ * @return `TILEDB_OK` for success and `TILEDB_OOM` or `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_alloc(
+    tiledb_ctx_t* ctx,
+    tiledb_array_t* array,
+    tiledb_layout_t layout,
+    tiledb_subarray_t** subarray);
+
+/**
+ * Deletes a subarray object.
+ *
+ * @param subarray The subarray object to be deleted.
+ */
+TILEDB_EXPORT void tiledb_subarray_free(tiledb_subarray_t** subarray);
+
+/**
+ * Retrieves the subarray layout.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_layout_t layout;
+ * tiledb_subarray_get_layout(ctx, subarray, &layout);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray.
+ * @param layout The layout to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_layout(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    tiledb_layout_t* layout);
+
+/**
+ * Retrieves the subarray type.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_datatype_t type;
+ * tiledb_subarray_get_type(ctx, subarray, &type);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray.
+ * @param type The type to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_type(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    tiledb_datatype_t* type);
+
+/**
+ * Retrieves the number of dimensions of a subarray.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * uint32_t dim_num;
+ * tiledb_subarray_get_ndim(ctx, subarray, &dim_num);
+ * @endcode
+ *
+ * @param ctx The TileDB context
+ * @param subarray The subarray.
+ * @param ndim The number of dimensions in the subarray.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_ndim(
+    tiledb_ctx_t* ctx, const tiledb_subarray_t* subarray, uint32_t* ndim);
+
+/**
+ * Retrieves the domain the subarray is constructed from.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * const void* domain;
+ * tiledb_subarray_get_domain(ctx, subarray, &domain);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param subarray The subarray.
+ * @param domain The domain to be retrieved.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT int32_t tiledb_subarray_get_domain(
+    tiledb_ctx_t* ctx, const tiledb_subarray_t* subarray, const void** domain);
```

Comment:
If we keep this, better would be to take a `tiledb_domain_t**` that is constructed for the user (and user must free), similar to the other `_get_*` APIs we have.